src/Store/FilterSubscription.re

```diff
@@ -0,0 +1,118 @@
+
+module Make = (JobConfig: Oni_Model.FilterJob.Config) => {
+  module Core = Oni_Core;
+  module Model = Oni_Model;
+  module Log = Core.Log;
+
+  module Actions = Model.Actions;
+  module Subscription = Core.Subscription;
+  module Job = Core.Job;
+
+  module FilterJob = Model.FilterJob.Make(JobConfig);
+
+  module Provider = {
+    type action = Actions.t;
+
+    type params = {
+      query: string,
+      items: list(JobConfig.item),
+      itemStream: Isolinear.Stream.t(list(JobConfig.item)),
+      onUpdate: (list(Model.Filter.result(JobConfig.item)), ~progress: float) => action
+    };
+
+    type state = {
+      job: FilterJob.t,
+      dispose: unit => unit
+    }
+
+    let jobs = Hashtbl.create(10);
+
+    let start = (~id, ~params as {query, items, itemStream, onUpdate}, ~dispatch) => {
+      Log.debug("Starting MenuJob subscription " ++ id);
+      let job =
+        FilterJob.create()
+          |> Job.map(FilterJob.updateQuery(query))
+          |> Job.map(FilterJob.addItems(items))
+          |> Job.doWork;
+
+      let unsubscribeFromItemStream =
+        Isolinear.Stream.subscribe(itemStream, items =>
+          switch (Hashtbl.find_opt(jobs, id)) {
+            | Some({ job } as state) =>
+              let job = Job.map(FilterJob.addItems(items), job) |> Job.doWork;
+              Hashtbl.replace(jobs, id, { ...state, job });
+
+            | None =>
+              Log.error("Unable to add items to non-existing FilterJob");
+          },
+        );
+
+      let disposeTick = 
+        Revery.Tick.interval(
+          _ =>
+            switch (Hashtbl.find_opt(jobs, id)) {
+              | Some({ job } as state) =>
+                if (!Job.isComplete(job)) {
+                  Hashtbl.replace(jobs, id, { ...state, job: Job.tick(job) });
+                }
+
+              | None =>
+                Log.error("Unable to tick non-existing FilterJob");
+            },
+          Seconds(0.),
+        );
+
+      let disposeMessagePump = 
+        Revery.Tick.interval(
+          _ =>
+            switch (Hashtbl.find_opt(jobs, id)) {
+              | Some({ job }) =>
+                let items =
+                  Job.getCompletedWork(job).uiFiltered
+                let progress =
+                  Job.getProgress(job);
+                dispatch(onUpdate(items, progress));
+
+              | None =>
+                Log.error("Unable to pump non-existing FilterJob");
+            },
+          Seconds(0.1),
+        );
+
+      let dispose = () => {
+        unsubscribeFromItemStream();
+        disposeTick();
+        disposeMessagePump();
+      };
+
+      Hashtbl.add(jobs, id, {job, dispose});
+    };
+
+    let update = (~id, ~params as {query}, ~dispatch) =>
+      switch (Hashtbl.find_opt(jobs, id)) {
+        | Some({ job } as state) =>
+          // `MenuJob.updateQuery` checks if `query` has changed, so we don't need to. It would make the commented out log message below less annoying though
```

Comment:
I believe this actually might be necessary, as if I check the `query`, the job actually gets completed: ```
          if (!String.equal(Job.getPendingWork(job).filter, query)) {
          let job = Job.map(FilterJob.updateQuery(query), job) |> Job.doWork;
          Hashtbl.replace(jobs, id, { ...state, job });
          }
```

One other thought - `FilterJob.updateQuery` _always resets the completed flag to `false`_ - it looks like we're calling this even if the job is complete... I think that's really the root cause. But just wrapping this in a block like this:
```
if (!Job.isComplete(job)) {
          let job = Job.map(FilterJob.updateQuery(query), job) |> Job.doWork;
          Hashtbl.replace(jobs, id, { ...state, job });
}
```

isn't the right fix - because we do want to reset the completion state to `false` if the query changed (for example - if I backspace - it'll need to filter a new set of results)

Reply:
Yeah, that was exactly it. An assumption made based on an oversimplified impression of `updateQuery` that was then forgotten about... :/ Fixed by actually checking whether the query has changed.

Thanks!

It also seems like I can now revert https://github.com/onivim/oni2/pull/858/commits/6276a8306d5bdf36d72dce6a0c72a92d2134d0e8 without strange things happening!

Reply:
Woohoo! Glad that was the issue 👍 I suspect that was maybe related to the weirdness described here too:
>  I also removed a few calls to Job.doWork that seemed unnecessary, but then weird(er) things started happening. 
(Since the `Job.doWork` would flip it back to completed). Anyway great news!